In [1]:
!pip install psycopg2-binary

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 12.3 MB/s eta 0:00:00a 0:00:01


In [3]:
import pandas as pd
import psycopg2

In [5]:
# DB connection
conn = psycopg2.connect(
    dbname = "neondb",
    user = "neondb_owner",
    password = "npg_CeS9fJg2azZD",
    host = "ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port = "5432",
    sslmode ="require"
)
cur = conn.cursor()

In [19]:
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

db_user = "your _username"
db_pass = "your_password"
db_host = "locahost"
db_port = "5432"
db_name = "training"

# SQLalchemy connection string
engine = create_engine(f"postgresql+psycopg2://{db_user}:{db_pass}@{db_host}:{db_port}/{db_name}")

In [29]:
from sqlalchemy import create_engine
import pandas as pd

# build engine string
db_user = "neondb_owner"
db_pass = "npg_CeS9fJg2azZD"
db_host = "ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech"
db_port = "5432"
db_name = "neondb"

engine = create_engine(
    f"postgresql+psycopg2://{db_user}:{db_pass}@{db_host}:{db_port}/{db_name}?sslmode=require"
)

# test query
query = "SELECT * FROM nyc_schools.high_school_directory LIMIT 10;"
df = pd.read_sql(query, engine)
df.head()

dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

   expgrade_span_min  expgrade_span_max  ... number_programs  \
0                NaN                NaN  ...               1   
1                NaN                NaN  ...               1   
2                NaN                NaN  ...               1   
3                NaN                NaN  ...               1   
4                NaN                NaN  ...               6   

                                          Location 1 Community Board  \
0  {'latitude': '40.601989336', 'longitude': '-73...              14   
1  {'latitude': '40.593593811', 'longitude': '-73...              13   
2  {'latitude': '40.692133704', 'longitude': '-73...               3   
3  {'latitude': '40.822303765', 'longitude': '-73...               9   
4  {'latitude': '40.773670507', 'longitude': '-73...               7   

  Council District Census Tract Zip Codes Community Districts  \
0               31       100802     20529                  51   
1               47          306     17616                  21   
2               36          291     18181                  69   
3               18           16     11611                  58   
4                6          151     12420                  20   

  Borough Boundaries City Council Districts Police Precincts  
0                  3                     47               59  
1                  2                     45               35  
2                  2                     49               52  
3                  5                     31               26  
4                  4                     19               12  

[5 rows x 105 columns]

Number of schools per borough

In [31]:
query = """
select borough, count(distinct dbn) AS total_schools
from nyc_schools.high_school_directory
group by borough
order by total_schools desc;
"""
df_schools = pd.read_sql(query, engine)
df_schools

borough  total_schools
0       Brooklyn            121
1          Bronx            118
2      Manhattan            106
3         Queens             80
4  Staten Island             10

Brooklyn has the most schools, while Staten Island has the least
avg percentage of ELL per borough.

In [39]:
query = """
SELECT h.borough, ROUND(AVG(d.ell_percent)::numeric, 2) AS avg_ell_percent
FROM nyc_schools.school_demographics d
JOIN  nyc_schools.high_school_directory h
ON d.dbn = h.dbn
GROUP BY h.borough
ORDER BY avg_ell_percent desc;
"""
df_ell = pd.read_sql(query, engine)
df_ell

borough  avg_ell_percent
0  Manhattan             7.57

Manhattan has an avg of 7.6% of ELL, which suggest a relatively low concentration of ELL students compared to other boroughs.

In [43]:
query = """
SELECT h.borough, h.school_name, d.sped_percent
FROM nyc_schools.school_demographics d
JOIN nyc_schools.high_school_directory h
ON d.dbn = h.dbn
Where d.sped_percent IS NOT NULL
ORDER BY h.borough, d.sped_percent desc
LIMIT 15;
"""
df_sped = pd.read_sql(query, engine)
df_sped

borough                                    school_name  sped_percent
0   Manhattan                     East Side Community School          28.8
1   Manhattan                     East Side Community School          27.7
2   Manhattan                     East Side Community School          26.7
3   Manhattan                     East Side Community School          26.4
4   Manhattan                        Marta Valle High School          25.9
5   Manhattan                     East Side Community School          25.1
6   Manhattan  Henry Street School for International Studies          25.1
7   Manhattan  Henry Street School for International Studies          24.9
8   Manhattan                     East Side Community School          24.8
9   Manhattan                     East Side Community School          24.5
10  Manhattan                        Marta Valle High School          24.4
11  Manhattan  Henry Street School for International Studies          24.1
12  Manhattan  Henry Street School for International Studies          23.7
13  Manhattan                        Marta Valle High School          23.3
14  Manhattan            University Neighborhood High School          23.1

Top 3 schools per borough by special education percentage

In [55]:
query = """
WITH ranked_schools AS (
  SELECT 
     h.borough,
     h.school_name,
     d.sped_percent,
     ROW_NUMBER() OVER (
     PARTITION BY h.borough
     ORDER BY d.sped_percent desc
     ) AS rank
FROM nyc_schools.school_demographics d
JOIN nyc_schools.high_school_directory h
ON d.dbn = h.dbn
WHERE d.sped_percent IS NOT NULL
)
SELECT borough, school_name, sped_percent
FROM ranked_schools
WHERE rank <= 3
ORDER BY borough, rank;
"""
df_sped = pd.read_sql(query, engine)
df_sped

borough                 school_name  sped_percent
0  Manhattan  East Side Community School          28.8
1  Manhattan  East Side Community School          27.7
2  Manhattan  East Side Community School          26.7

The percentage of educational students ranging from 26-29% is considerably high in the east side community school in Manhattan
indicating its strong role in supporting students.

INSIGHTS FROM NYC SCHOOL DEMOGRAPHICS

- Brooklyn has the highest number of schools (121), followed by Bronx with (118) and Manhattan (106). Staten Island is the least as school distribution.
- East side community school appears to be the leader, highlighting its strong focus on inclusive education. 